Slim Models.py

JianyuZhan opened the pull request:
Since https://github.com/OpenNMT/OpenNMT-py/pull/230 is merged, we can slim `Models.py` a little bit.  Ideally, when the `Encoder` refactor work is done and `Embeddings` is moved to its own file.  `Models.py` should contain only these:

1. `make_base_model()`,  `make_embeddings()`, `make_encoder()`, and `make_decoder()`.
2. The basic RNN based Encoder/Decoder classes and their `*DecoderState`'s.

So, this PR just moves `Transformer` and `Conv2Conv` specific classes into their own files.

**Patch list**

Only **2** patches in this PR.

1. 82a07b3: Move `TransformerDecoder` and `TransformerDecoderState` to `onmt/modules/Transformer.py`, and rename original `onmt.modules.TransformerDecoder` to `onmt.modules.TransformerDecoderLayer`.  When this is done, there is a circular dependency between `Models.py` and `Transformer.py`(former  one need later's `TransformerDecoderState`,  later one needs former's `DecoderState`), so I put the `DecoderState` into `Translator.py`, because it is related to translation.

2. eaec0c8: Move `CNNDecoder` and `CNNDecoderState` to `onmt/modules/Conv2Conv.py`.  While at it,  collapse the `onmt.modules.ConvDecoder` into `CNNDecoder`, because it is unnecessary to add another wrapper.

srush commented on onmt/Models.py:
```diff
@@ -676,10 +482,7 @@ def init_decoder_state(self, src, context, enc_hidden):
-            scale_weight = 0.5 ** 0.5
-            init_state.init_src = (context + enc_hidden) * scale_weight
-            return init_state
+            return CNNDecoderState(context, enc_hidden)
```
Not sure where this method belongs, but probably should be associated with each decoder not NMT model. Can you use this PR to move these functions to each Decoder?

JianyuZhan replied:
Make sense. Will do.

srush commented on onmt/Models.py:
```diff
@@ -2,13 +2,18 @@
 from onmt.modules.Gate import ContextGateFactory
-from torch.nn.utils.rnn import pad_packed_sequence as unpack
-from torch.nn.utils.rnn import pack_padded_sequence as pack
+from onmt.Translator import DecoderState
```
No, I don't think this file / class should know about Translator, let's keep state here.

JianyuZhan replied:
Yes. The `DecoderState` should be moved to some other place.  Otherwise there would be a circular dependency(`Models.py` needs the `TransformerDecoderState` in `Transformer.py`,  while `Transformer.py` need `DecoderState` in `Models.py`).

srush commented on onmt/Translator.py:
```diff
@@ -6,6 +6,28 @@
 from torch.autograd import Variable
 
 
+class DecoderState(object):
```
I think this should be in model and not here, unfortunately. We need to think a bit more about the names of repeatBeam and beamUpdate.

JianyuZhan commented:
**v1->v2:**

V2 addressed the review feedback in v1, and fixed up some leftover in `Embeddings` refactor work.
See below for details.

**Patch  list**
Total **5** patches in v2.

1 and 2 are the same as v1.

3. 6445afc Embeddings: fix stale comments, naming inconsistency, and leftover change of `make_embeddings` in `test/test_models.py`.

4. 48b0d10 DecoderState: move init_decoder_state() to each *Decoder class.
   
To keep `DecoderState` in `Models.py`, we need to eschew the circular dependency problem between `Models.py` and `Transformer.py`.  The approach is to avoid
`from onmt.modules... import ....`  syntax, instead using full qualified names.  And this renders
that `aeq` should not stay in the `onmt.modules` package(it is used by upper layer package too), so move it to a new `onmt/Utils.py`, in which we later could put config-saving/loading stuff, log utils, etc.  And rename `onmt/modules/Util.py` to `onmt/modules/UtilClass.py`.

5. c488047 DecoderState: further refactor

This further refactor `*DecoderState` classes, including:
- Minimize each *DecoderState's contructor.
- Add an update_state() for each *DecoderState, when updating the state in the decoder forward pas
- Make all a property: _all(), so now we don't need to update it when update the attributes.
- Rename: `repeatBeam_` => `repeat_beam_size`, `beamUpdate_` => `beam_update`, with proper
    docstring added.

bpopeters commented:
I've also been working on making the Encoder more module and allowing it to reside in its own file in onmt/modules. Perhaps our efforts here are complementary.

srush commented:
I'm going to accept this as I think it improves things. 

@bpopeters and @JianyuZhan maybe we can setup a call sometime in the future, perhaps with some other contributors? It would be nice to have an explicit roadmap so you don't overlap in your incredible work. 

Couple more changes I would like in the future: 

- We should now split all the make_* functions out of model.py and into a `ModelConstructor.py` file which would be the only file that can look at `opts`. I really like now that you can build an NMT system from its parts directly and not have to use `translate.py`. 
- It should be called `CNNEncoder` as well 
- I'm realize now that I am not 100% happy with the class hierarchy of decoder. I think I would prefer each class had its own `forward` method. If you think they need to share code have the subclasses do it explicityl.
-

JianyuZhan commented:
@bpopeters , I am doing the `Encoder` split work now.  I think maybe I should post it early and you can do further work on it . I think I should first split them into several own classes, and make a `make_encoder()` function creating them.  The structure of them is clean and less complex than the `*Decoder` part, and as @srush prefer each has an explicit `forward()` function,  I think each should just inherits from `nn.Module`(I am doing it now, and found they really share little code,  but much more the shape).  And then I will move all `make*()` function to its own file.

Later, based on this, we could simplify and shrink their interfaces(`forward()`). 

As for the `Decoder`, I am also not satisfied with the hierarchy here - a little fragile.  It would be better to share a common base class, and put common code in it, then each one just inherits from it.  I will do
it later. Thoughts? Thanks.

bpopeters commented:
This is just a possible Encoder.py file that I wrote up during some downtime at EMNLP. I haven't tested it yet at all -- it's just an illustration of a style that I think could clarify what's going on. The `CNNEncoder` class can be removed if we incorporate the tranposing, etc. in `onmt.modules.ConvEncoder`, and the `TransformerEncoder` can be moved to Transformer.py. I renamed the class with the same name in Transformer.py to `TransformerEncoderLayer`, which I think better describes its purpose.

What I like about this way of doing it is that it leaves the forward pass of the encoder network as simple as possible and makes it easy to define new kinds of encoder without having to mess around with the basic forward pass. The Encoder also needs way fewer attributes.

```
import torch.nn as nn
from torch.autograd import Variable
import onmt
import onmt.modules
from onmt.modules import aeq
from torch.nn.utils.rnn import pad_packed_sequence as unpack
from torch.nn.utils.rnn import pack_padded_sequence as pack


class MeanEncoder(nn.Module):
    """
    Mean encoder: a small module for computing the mean of a sequence of
    embeddings.
    """
    def __init__(self, num_layers):
        self.layers = num_layers
        super(MeanEncoder, self).__init__()

    def forward(self, emb, **kwargs):
        """
        emb (FloatTensor): src_len x batch x embedding dimension
        returns:
        """
        _, n_batch, emb_dim = emb.size()
        mean = emb.mean(0).expand(self.layers, n_batch, emb_dim)
        return (mean, mean), emb


class TransformerEncoder(nn.Module):
    """
    Transformer encoder from "Attention is All You Need"
    for discussion: Move to Transformer.py?
    """
    def __init__(self, hidden_size, dropout, num_layers, padding_idx):
        self.padding_idx = padding_idx
        super(TransformerEncoder, self).__init__()
        self.transformer = nn.ModuleList(
                [onmt.modules.TransformerEncoderLayer(hidden_size, dropout)
                 for i in range(num_layers)])

    def forward(self, emb, input, **kwargs):
        """
        emb: src_len x batch x embedding dimension
        input (LongTensor): src_len x batch x nfeat
        """
        out = emb.transpose(0, 1).contiguous()
        words = input[:, :, 0].transpose(0, 1)
        # CHECKS
        out_batch, out_len, _ = out.size()
        w_batch, w_len = words.size()
        aeq(out_batch, w_batch)
        aeq(out_len, w_len)
        # END CHECKS

        # Make mask.
        mask = words.data.eq(self.padding_idx).unsqueeze(1) \
            .expand(w_batch, w_len, w_len)

        # Run the forward pass of every layer of the tranformer.
        for tf in self.transformer:
            out = tf(out, mask)

        # I don't know how Variable(emb.data) varies from emb
        return Variable(emb.data), out.transpose(0, 1).contiguous()


class CNNEncoder(nn.Module):
    """
    Convolutional encoder
    Question: why isn't this behavior implemented in onmt.modules.ConvEncoder?
    """
    def __init__(self, embedding_dim, hidden_size, dropout,
                 layers, kernel_width):
        super(CNNEncoder, self).__init__()
        self.cnn = onmt.modules.ConvEncoder(
                embedding_dim, hidden_size,
                layers, dropout, kernel_width)

    def forward(self, emb, **kwargs):
        out = emb.transpose(0, 1).contiguous()
        out, emb_remap = self.cnn(out)
        return (emb_remap.transpose(0, 1).contiguous(),
                out.transpose(0, 1).contiguous())


class RNNEncoder(nn.Module):
    """
    Standard RNN encoder. Handles layer sizing issues for bidirectional
    networks and allows for sequence packing.
    """
    def __init__(self, rnn_type, embedding_dim, rnn_size,
                 num_layers, dropout, bidirectional):
        assert rnn_type in ['LSTM', 'GRU']
        # TODO: implement sum BRNN merge
        num_directions = 2 if bidirectional else 1
        assert rnn_size % num_directions == 0
        hidden_size = rnn_size // num_directions
        super(RNNEncoder, self).__init__()
        self.rnn = getattr(nn, rnn_type)(
            input_size=embedding_dim,
            hidden_size=hidden_size,
            num_layers=num_layers,
            dropout=dropout,
            bidirectional=bidirectional)

    def forward(self, emb, lengths=None, hidden=None, **kwargs):
        packed_emb = emb
        if lengths is not None:
            # Lengths data is wrapped inside a Variable.
            lengths = lengths.view(-1).tolist()
            packed_emb = pack(emb, lengths)
        outputs, hidden_t = self.rnn(packed_emb, hidden)
        if lengths:
            outputs = unpack(outputs)[0]
        return hidden_t, outputs


class Encoder(nn.Module):
    """
    Encoder recurrent neural network. The encoder consists of two basic
    parts: an embedding matrix for mapping input tokens to fixed-dimensional
    vectors, and a unit which processes these embedded tokens.
    """

    def __init__(self, encoder_type, bidirectional, rnn_type,
                 num_layers, rnn_size, dropout, embeddings,
                 cnn_kernel_width=3):
        """
        Args:
            encoder_type (string): rnn, brnn, mean, or transformer.
            bidirectional (bool): bidirectional Encoder.
            rnn_type (string): LSTM or GRU.
            num_layers (int): number of Encoder layers.
            rnn_size (int): size of hidden states of a rnn.
            dropout (float): dropout probablity.
            embeddings (Embeddings): vocab embeddings for this Encoder.
        """
        super(Encoder, self).__init__()

        self.embeddings = embeddings

        # Build the Encoder RNN.
        if self.encoder_type == "mean":
            self.encoder = MeanEncoder()
        elif self.encoder_type == "transformer":
            self.encoder = TransformerEncoder(
                rnn_size, dropout, num_layers,
                self.embeddings.padding_idx)
        elif self.encoder_type == "cnn":
            self.encoder = CNNEncoder(
                self.embeddings.embedding_dim, rnn_size,
                self.num_layers, dropout, cnn_kernel_width)
        else:
            self.encoder = RNNEncoder(
                rnn_type, self.embeddings.embedding_dim, rnn_size,
                num_layers, dropout, bidirectional)

    def forward(self, input, lengths=None, hidden=None):
        """
        Args:
            input (LongTensor): len x batch x nfeat
            lengths (LongTensor): batch
            hidden: Initial hidden state.
        Returns:
            hidden_t (FloatTensor): Pair of layers x batch x rnn_size - final
                                    Encoder state
            outputs (FloatTensor):  len x batch x rnn_size -  Memory bank
        """
        # CHECKS
        s_len, n_batch, n_feats = input.size()
        if lengths is not None:
            n_batch_, = lengths.size()
            aeq(n_batch, n_batch_)
        # END CHECKS

        emb = self.embeddings(input)

        return self.encoder(emb, input=input, lengths=lengths, hidden=hidden)

```

JianyuZhan commented:
This is almost the same with my code, except the final `Encoder` class. To keep the symmetry between encoder and decoder side, I also make it a dispatch function named `make_encoder()`,  and at first I also implement a base class `EncoderBase`, whose `forward()` acts as the `forward()` in your `Encoder()`. But since @srush prefers a explicit `forward()` for each `*Encoder`, I now make each class implement its own `forward()`.
